sdk/eventhub/event-hubs/src/eventHubReceiver.ts

```diff
@@ -200,214 +165,234 @@ export class EventHubReceiver extends LinkEntity {
+    if (this._hasActiveListeners() && !this._isDrainingQueue) {
+      this._addCredit(Math.max(this.prefetchCount - this._existingCredits(), 1));
+    }
 
-    this._onSessionClose = async (context: EventContext) => {
-      const receiver = this._receiver || context.receiver!;
-      const sessionError = context.session && context.session.error;
-      if (sessionError) {
-        log.error(
-          "[%s] 'session_close' event occurred for receiver '%s' with address '%s'. " +
-            "The associated error is: %O",
-          this._context.connectionId,
-          this.name,
-          this.address,
-          sessionError
-        );
-      }
+    if (this._userDefinedOnMessage && !this._isDrainingQueue) {
+      this._userDefinedOnMessage(receivedEventData);
+    } else {
+      this.internalQueue.push(receivedEventData);
+    }
+  }
 
-      if (receiver && !receiver.isSessionItselfClosed()) {
-        if (!this.isConnecting) {
-          log.error(
-            "[%s] 'session_close' event occurred on the session of receiver '%s' with " +
-              "address '%s' and the sdk did not initiate this. Hence calling detached from the " +
-              "_onSessionClose() handler.",
-            this._context.connectionId,
-            this.name,
-            this.address
-          );
-          await this.onDetached(sessionError);
-        } else {
-          log.error(
-            "[%s] 'session_close' event occurred on the session of receiver '%s' with " +
-              "address '%s' and the sdk did not initiate this. Moreover the receiver is already " +
-              "re-connecting. Hence not calling detached from the _onSessionClose() handler.",
-            this._context.connectionId,
-            this.name,
-            this.address
-          );
-        }
-      } else {
-        if (this._abortSignal) {
-          this._abortSignal.removeEventListener("abort", this._onAbort);
-        }
+  private _onAmqpError(context: EventContext): void {
+    const amqpReceiver = this._receiver || context.receiver;
+    if (!amqpReceiver) {
+      return;
+    }
+
+    const amqpError = amqpReceiver.error;
+    if (!amqpError) {
+      return;
+    }
+
+    const error = translate(amqpError);
+    log.error(
+      "[%s] An error occurred for Receiver '%s': %O.",
+      this._context.connectionId,
+      this.name,
+      error
+    );
+
+    if (!amqpReceiver.isItselfClosed() && this._userDefinedOnError) {
+      log.error(
+        "[%s] Since the user did not close the receiver " +
+          "we let the user know about it by calling the user's error handler.",
+        this._context.connectionId
+      );
+      this._userDefinedOnError(error);
+    } else {
+      log.error(
+        "[%s] Tthe receiver was closed by the user." +
+          "Hence not notifying the user's error handler.",
+        this._context.connectionId
+      );
+    }
+  }
+
+  private async _onAmqpSessionClose(context: EventContext): Promise<void> {
```

Comment:
Please re-order with either
- `_onAmqpError`
-  `_onAmqpClose`
- `_onAmqpSessionError`
-  `_onAmqpSessionClose`

or

- `_onAmqpError`
- `_onAmqpSessionError`
-  `_onAmqpClose`
-  `_onAmqpSessionClose`

for ease of readability